Script style end parsing

pngwn opened the pull request:
This updates the parsing of script and style end tags to be more compliant with the HTML spec. The spec allows for arbitrary whitespace of almost any kind and any length after the tagName and before the closing `>`.

I had to modify the `read_until` method to return the actual matched string as well as the the string up to the match, as we can no longer predict that ahead of time. We need the actual string so we can `eat` it. The script end tag parsing works differently (ish) so it was not affected by this change. They both now use `regex.exec` to match, so it might be easy enough to cleanup those differences at some point.

I also added a couple of new tests to make sure things work but mostly to check that the positional data is correct.

tanhauhau commented on src/compiler/parse/read/script.ts:
```diff
@@ -30,17 +28,17 @@ function get_context(parser: Parser, attributes: any[], start: number): string {
-	const script_end = parser.template.indexOf(script_closing_tag, script_start);
-
-	if (script_end === -1) parser.error({
+	const script_end = /<\/script\s*>/.exec(parser.template.slice(script_start));
```
use `parser.read_until` here instead?

pngwn commented:
I've updated the `read_script` util to use the `parser.read_until` method which broke the nice error messages in the case of no closing script (because `read_until` throws early if there is an unexpected `eof`). `eof` errors are a good fallback but there is nothing actionable in the message, even though we know what the problem is. I have modified `read_until` to accept an optional error message to display in its place to solve this problem.

I've added some additional tests to check the error message for no closing tags in the eof case and the non-eof case.

This does mean that errors are handled in two places, `eof` errors are handled inside the `read_until` method and match failures (when no unexpected eof) are handled in the `read_style` and `read_script` utilities. 

I feel like errors should probably generally be handled outside of the `read_until` util), sometimes you can't really provide actionable information and an `unexpected eof` is reasonable. At other times a single error message serves best in either case. only the caller will really know what is best. I haven't made these changes yet because I'm wary of inconsistent returns, this util can already has two return types.

pngwn commented:
This would probably easier to match against:
```ts
// data could be an object containing the substring and the match 
type success = [ true, data ] 
// data could be an object containing a reason for the failure (if needed) and the sliced string as it is now in failed matches
type failure = [ false, data ] 
```

Then you could to something like this as the caller:
```ts
const [ success, data ] = parser.read_until(regex);

if (!success) {
  if (data.reason === 'eof') parser.error(something);
  else return null;
}

const ast = parse(data.source);
parser.eat(data.match);
```

pngwn commented:
I'm going to do some more work on the error handling here, it is a bit problematic just changing `read_until`. I've marked this as a draft for now and will have some more in a few days.

Let reconsider this PR to be 'refactor error handling to be a bit more consistent'.

pngwn commented:
forgot about this
